core/src/main/java/org/apache/calcite/sql/fun/SqlCaseOperator.java

```diff
@@ -230,14 +230,25 @@ private RelDataType inferTypeFromValidator(
     SqlNodeList thenList = caseCall.getThenOperands();
     ArrayList<SqlNode> nullList = new ArrayList<>();
     List<RelDataType> argTypes = new ArrayList<>();
-    for (SqlNode node : thenList) {
-      argTypes.add(
-          callBinding.getValidator().deriveType(
-              callBinding.getScope(), node));
+
+    final SqlNodeList whenOperands = caseCall.getWhenOperands();
+    final RelDataTypeFactory typeFactory = callBinding.getTypeFactory();
+
+    final int size = thenList.getList().size();
+    for (int i = 0; i < size; i++) {
+      SqlNode node = thenList.get(i);
+      RelDataType type = callBinding.getValidator().deriveType(
+          callBinding.getScope(), node);
+      // We're sure that the type is not nullable if its kind is IS NOT NULL.
+      if (whenOperands.get(i).getKind() == SqlKind.IS_NOT_NULL && type.isNullable()) {
+        type = typeFactory.createTypeWithNullability(type, false);
```

Comment:
I think we can also give the nullability fix for `IS_NULL`.

Reply:
Make sense. Will do. Thanks for your suggestion.

Reply:
I just find this fix is wrong. For instance, 
for `case when name is not null then age else 20`, it should be nullable when age is nullable.
I will spend more time on it.